src/include/platform/NetworkCommissioning.h

```diff
@@ -0,0 +1,260 @@
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines the CHIP Device Network Provisioning object.
+ *
+ */
+
+#pragma once
+
+#include <lib/core/CHIPCore.h>
+#include <lib/core/CHIPSafeCasts.h>
+#include <lib/support/ThreadOperationalDataset.h>
+#include <platform/internal/DeviceNetworkInfo.h>
+
+namespace chip {
+namespace DeviceLayer {
+/**
+ * We are using a namespace here, for most use cases, this namespace will be used by `using DeviceLayer::NetworkCommissioning`, but
+ * this still worth a dedicated namespace since:
+ *
+ * - The BaseDriver / WirelessDriver is not expected to be implemented directly by users, the only occurrance is in the network
+ *   commissioning cluster.
+ * - We can safely name the drivers as WiFiDriver / ThreadDriver, it should not be ambiguous for most cases
+ * - We can safely name the Status enum to Status, and some other structs -- if we are using using, then we should in the context of
+ *   writing something dedicated to network commissioning, then a single word Status should be clear enough.
+ */
+namespace NetworkCommissioning {
+
+constexpr size_t kMaxNetworkIDLen = 32;
+
+// TODO: This is exactly the same as the one in GroupDataProvider, this could be moved to src/lib/support
+template <typename T>
+class Iterator
+{
+public:
+    virtual ~Iterator() = default;
+    /**
+     *  @retval The number of entries in total that will be iterated.
+     */
+    virtual size_t Count() = 0;
+    /**
+     *   @param[out] item  Value associated with the next element in the iteration.
+     *  @retval true if the next entry is successfully retrieved.
+     *  @retval false if no more entries can be found.
+     */
+    virtual bool Next(T & item) = 0;
+    /**
+     * Release the memory allocated by this iterator.
+     * Must be called before the pointer goes out of scope.
+     */
+    virtual void Release() = 0;
+
+protected:
+    Iterator() = default;
+};
+
+/**
+ * The content should match the one in zap_generated/cluster-objects.h
+ */
+enum class Status : uint8_t
+{
+    kSuccess                = 0x00,
+    kOutOfRange             = 0x01,
+    kBoundsExceeded         = 0x02,
+    kNetworkIDNotFound      = 0x03,
+    kDuplicateNetworkID     = 0x04,
+    kNetworkNotFound        = 0x05,
+    kRegulatoryError        = 0x06,
+    kAuthFailure            = 0x07,
+    kUnsupportedSecurity    = 0x08,
+    kOtherConnectionFailure = 0x09,
+    kIPV6Failed             = 0x0A,
+    kIPBindFailed           = 0x0B,
+    kUnknownError           = 0x0C,
+};
+
+// The following structs follows the generated cluster object structs.
+struct Network
+{
+    uint8_t networkID[kMaxNetworkIDLen];
+    uint8_t networkIDLen;
+    bool connected;
+};
+
+struct WiFiScanResponse
+{
+public:
+    uint8_t security;
+    uint8_t ssid[DeviceLayer::Internal::kMaxWiFiSSIDLength];
+    uint8_t ssidLen;
+    uint8_t bssid[6];
+    uint8_t channel;
+    uint32_t wiFiBand;
```

Comment:
This could be replaced by a 8 bit enum. The spec now uses a WifiBand enum:

```
=== WiFiBand Enum
The `WiFiBand` type is derived from the <<enum-def,enum8>> data type.

The `WiFiBand` enum encodes a supported Wi-Fi frequency band present in the `WiFiBand` field of the <<ref_WiFiInterfaceScanResult_Struct,`WiFiInterfaceScanResult`>>.

[options="header",valign="middle"]
|===
s| Value | Description
s| 0     | 2.4GHz - 2.401GHz to 2.495GHz (802.11b/g/n/ax)
s| 1     | 3.65GHz - 3.655GHz to 3.695GHz (802.11y)
s| 2     | 5GHz - 5.150GHz to 5.895GHz (802.11a/n/ac/ax)
s| 3     | 6GHz - 5.925GHz to 7.125GHz (802.11ax / WiFi 6E)
s| 4     | 60GHz - 57.24GHz to 70.20GHz (802.11ad/ay)
|===
```